src/libfuzzer/libfuzzer_mutator.cc

```diff
@@ -22,8 +22,7 @@
 #include "port/protobuf.h"
 #include "src/mutator.h"
 
-extern "C" size_t LLVMFuzzerMutate(uint8_t*, size_t, size_t)
-    __attribute__((weak));
+extern "C" size_t LLVMFuzzerMutate(uint8_t*, size_t, size_t);
```

Comment:
LibFuzzer on MSVC doesn't use `__attribute__((weak))` (see [here](https://github.com/llvm-mirror/compiler-rt/blob/d798ef3964f20a6b76e513637c965ddeb8f7b3a0/lib/fuzzer/FuzzerExtFunctionsWindows.cpp#L35)). Either way, I don't think we need to declare this as a weak symbol for clang builds to be happy.

Reply:
As-is it's going to break use-case when we test corpus without mutation.
In this case we don't link libfuzzer so symbol is missing. Should be true for windows as well.
We can avoid that if we split lib into parts where one part does not nee libfuzzer and another needs. But it will require some efforts.

Reply:
Can you elaborate? With this change, all tests seem to build and pass on linux (tests fail to compile on Windows for other reasons, such as cmake failing to find libgtest.a--it's gtest.lib on Windows).

Tests link only against protobuf-mutator, zlib and gtest, so they don't seem to link against libfuzzer.

Reply:
Unfortunately this case is not tested here, but it should be easy to test.

Let consider non libprotobuf-mutator case. You have a fuzzer e.g.
```
// fuzz_target.cc
extern "C" int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
  DoSomethingInterestingWithMyAPI(Data, Size);
  return 0;  // Non-zero return values are reserved for future use.
}
```

Usually you link it with libFuzzer and libFuzzer calls it.
Then you found interesting input which crashed something. When bug is fixed you may want to have some regression test for that e.g.:

```
// regression_test.cc
int main() {
  LLVMFuzzerTestOneInput("bad data", size);
}
```
in this case you don't need to link it with libFuzzer

Now lets get back to libprotobuf-mutator.
```
DEFINE_PROTO_FUZZER(const libfuzzer_example::Msg& message) {
    DoSomethingInterestingWithMyMessageI(message);
}
```
DEFINE_PROTO_FUZZER creates:
1. LLVMFuzzerTestOneInput which just parses a message.
2. LLVMFuzzerCustomMutator/Crossover which mutate a message

Our regression tests are still very similar:
```
int main() {
  LLVMFuzzerTestOneInput("bad serialized message", size);
}
```
It calls LLVMFuzzerTestOneInput from DEFINE_PROTO_FUZZER, but there is unneeded LLVMFuzzerCustomMutator/Crossover which need LLVMFuzzerMutate.

This is how we run regression tests inside of Google, Chromium should be setup same way.
main() is more complicated, It loads inputs from file but idea is the same.

So how to avoid that without weak?
I guess we can just link dummy LLVMFuzzerMutate if we don't link libfuzzer.
Let me try this on our code, if it works we will merge the patch as-is and I cover this case with a test or/and documentation.

Reply:
If this doesn't work, I can put back the `__attribute__((weak))` and put it behind an ifdef. I can also test this exact scenario out on Windows this weekend; I suspect something like this might work:

```c
#ifdef LIBFUZZER_MSVC
  extern "C" size_t LLVMFuzzerMutateDef(uint8_t*, size_t, size_t);
  __pragma(comment(linker, "/alternatename:LLVMFuzzerMutate=LLVMFuzzerMutateDef"));
#else
  extern "C" size_t LLVMFuzzerMutate(uint8_t*, size_t, size_t)
    __attribute__((weak));
#endif
```

Reply:
```
#ifdef MSVC
# define ATTRIBUTE_WEAK ;__pragma(comment(linker, "/alternatename:LLVMFuzzerMutate=LLVMFuzzerMutateDef"));
#else 
# define ATTRIBUTE_WEAK __attribute__((weak))
#endif 
```

Reply:
Maybe something like this?
https://github.com/llvm/llvm-project/blob/885b790f89b6068ec4caad8eaa51aa8098327059/compiler-rt/lib/sanitizer_common/sanitizer_internal_defs.h#L56

Reply:
Hmm, that looks to eventually do the same thing:

```
#define WIN_WEAK_ALIAS(Name, Default)                                          \
  __pragma(comment(linker, "/alternatename:" WIN_SYM_PREFIX STRINGIFY(Name) "="\
                                             WIN_SYM_PREFIX STRINGIFY(Default)))
```

Reply:
Yes, my point is that we can't just define ATTRIBUTE_WEAK and we need macro with name argument

Reply:
Would you like me to update this PR with a change for this?

Reply:
yes please

Reply:
Is there anything more that should be done for this?